zaza/openstack/charm_tests/heat/tests.py

```diff
@@ -0,0 +1,311 @@
+        env_files, env = template_utils.process_environment_and_files(
+            env_path=None)
+
+        fields = {
+            'stack_name': STACK_NAME,
+            'timeout_mins': '15',
+            'disable_rollback': False,
+            'parameters': {
+                'admin_pass': 'Ubuntu',
+                'key_name': nova_utils.KEYPAIR_NAME,
+                'image': IMAGE_NAME
+            },
+            'template': template,
+            'files': dict(list(t_files.items()) + list(env_files.items())),
+            'environment': env
+        }
+
+        # Create the stack
+        try:
+            stack = self.heat_client.stacks.create(**fields)
+            logging.info('Stack data: {}'.format(stack))
+            stack_id = stack['stack']['id']
+            logging.info('Creating new stack, ID: {}'.format(stack_id))
+        except Exception as e:
+            # Generally, an api or cloud config error if this is hit.
+            msg = 'Failed to create heat stack: {}'.format(e)
+            logging.error(msg)
+            raise
+
+        # Confirm stack reaches COMPLETE status.
+        # /!\ Heat stacks reach a COMPLETE status even when nova cannot
+        # find resources (a valid hypervisor) to fit the instance, in
+        # which case the heat stack self-deletes!  Confirm anyway...
+        openstack_utils.resource_reaches_status(self.heat_client.stacks,
+                                                stack_id,
+                                                expected_status="COMPLETE",
+                                                msg="Stack status wait")
+        # List stack
+        stacks = list(self.heat_client.stacks.list())
+        logging.info('All stacks: {}'.format(stacks))
+
+        # Get stack information
+        try:
+            stack = self.heat_client.stacks.get(STACK_NAME)
+        except Exception as e:
+            # Generally, a resource availability issue if this is hit.
+            msg = 'Failed to get heat stack: {}'.format(e)
+            assert False, msg
+
+        # Confirm stack name.
+        logging.info('Expected, actual stack name: {}, '
+                     '{}'.format(STACK_NAME, stack.stack_name))
+        if STACK_NAME != stack.stack_name:
+            msg = 'Stack name mismatch, {} != {}'.format(STACK_NAME,
+                                                         stack.stack_name)
+            assert False, msg
+
+        # Confirm existence of a heat-generated nova compute resource
+        logging.info('Confirming heat stack resource status...')
+        resource = self.heat_client.resources.get(STACK_NAME, RESOURCE_TYPE)
+        server_id = resource.physical_resource_id
+        if server_id:
+            logging.debug('Heat template spawned nova instance, '
+                          'ID: {}'.format(server_id))
+        else:
+            msg = 'Stack failed to spawn a nova compute resource (instance).'
+            logging.error(msg)
+
```

Comment:
Either an error has occurred and the test should bail out or it hasn't and there is nothing to log?  Which is it?  I suspect this should just be:

```python
self.assertTrue(server_id, "Stack failed to spawn a nova compute ...")
```